Added dockerfile for rosbag use and updated readme.md for ros2 use.

Udaykiran87 opened the pull request:
Added new dockerfile in case user wants to use rosbag file for running openvslam.
Added detailed steps to run openvslam using ros2 in docker environment.

ymd-stella commented:
Please reconsider with the approach of creating another container to play rosbag. Maintaining Dockerfile.socket-rosbag is very annoying. Please remove unnecessary dependencies to make it maintainable. Since OpenVSLAM and rosbag do not have any dependencies on each other, it should be possible to separate them into two containers.

ymd-stella commented:
Similarly for image-tools. Avoid installing unnecessary packages for those who do not need them. In particular, there are many camera driver nodes. For example, https://github.com/clydemcqueen/gscam2, https://github.com/IntelRealSense/realsense-ros, etc.   
It would be ridiculous to consider all of them. It is fine to give examples of using a particular camera driver, but if you create a Dockerfile for it, it is unlikely that other users will be able to use it as-is.

Udaykiran87 commented:
> rosbag

I understood. Thanks for your feedback.

Udaykiran87 commented:
> Similarly for image-tools. Avoid installing unnecessary packages for those who do not need them. In particular, there are many camera driver nodes. For example, https://github.com/clydemcqueen/gscam2, https://github.com/IntelRealSense/realsense-ros, etc.
> It would be ridiculous to consider all of them. It is fine to give examples of using a particular camera driver, but if you create a Dockerfile for it, it is unlikely that other users will be able to use it as-is.

I understood. Thanks for your feedback.

ymd-stella commented:
> I am sorry, just got confused. Do you suggest me to edit and rename Dockerfile.socket-rosbag? Do you want me to create another container just like running a localhost server? Could you please simplify " remove unnecessary dependencies to make it maintainable".

Container A: Container created with existing Dockerfile.socket
Container B: A new container with rosbag installed

Udaykiran87 commented:
> > I am sorry, just got confused. Do you suggest me to edit and rename Dockerfile.socket-rosbag? Do you want me to create another container just like running a localhost server? Could you please simplify " remove unnecessary dependencies to make it maintainable".
> 
> Container A: Container created with existing Dockerfile.socket
> Container B: A new container with rosbag installed

@ymd-stella 
I have updated readme.md as per your review comments. I have deleted old Dockerfile.socket-rosbag and added new Dockerfile.rosbag. Could you please review it again.

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+    ros-${ROS2_DISTRO}-image-transport \
+    ros-${ROS2_DISTRO}-cv-bridge \
+    ros-${ROS2_DISTRO}-image-tools \
+    ros-foxy-rosbag2-bag-v2-plugins \
+    ros-foxy-ros2bag \
+    ros-foxy-rosbag2-transport \
+    ros-foxy-rosbag2-converter-default-plugins \
+    ros-foxy-rosbag2-storage-default-plugins \
```
Do not mix up foxy and galactic.

```suggestion
    ros-${ROS2_DISTRO}-rosbag2-bag-v2-plugins \
    ros-${ROS2_DISTRO}-ros2bag \
    ros-${ROS2_DISTRO}-rosbag2-transport \
    ros-${ROS2_DISTRO}-rosbag2-converter-default-plugins \
    ros-${ROS2_DISTRO}-rosbag2-storage-default-plugins \
```

Udaykiran87 replied:
@ymd-stella 
Thank you so much for your review comments. After updating all changes as suggested by you, I got bellow error while docker build.
`E: Unable to locate package ros-galactic-rosbag2-converter-default-plugins`

So just as a work around i replaced   `${ROS2_DISTRO}` with `foxy` in `ros-${ROS2_DISTRO}-rosbag2-converter-default-plugins`. This time build was successful. But i was unable to play rosbag. Got below error.

```
Failed to load entry point 'play': /opt/ros/galactic/lib/librosbag2_storage.so: undefined symbol: _ZNK12class_loader11ClassLoader15isLibraryLoadedEv
Traceback (most recent call last):
  File "/opt/ros/galactic/bin/ros2", line 11, in <module>
    load_entry_point('ros2cli==0.13.1', 'console_scripts', 'ros2')()
  File "/opt/ros/galactic/lib/python3.8/site-packages/ros2cli/cli.py", line 39, in main
    add_subparsers_on_demand(
  File "/opt/ros/galactic/lib/python3.8/site-packages/ros2cli/command/__init__.py", line 250, in add_subparsers_on_demand
    extension.add_arguments(
  File "/opt/ros/galactic/lib/python3.8/site-packages/ros2bag/command/bag.py", line 26, in add_arguments
    add_subparsers_on_demand(
  File "/opt/ros/galactic/lib/python3.8/site-packages/ros2cli/command/__init__.py", line 237, in add_subparsers_on_demand
    extension = command_extensions[name]
KeyError: 'play'
```
Commands followed are as below:
root@hostname:/ros2_ws#source /opt/ros/${ROS1_DISTRO}/setup.bash
root@hostname:/ros2_ws#source /opt/ros/${ROS2_DISTRO}/setup.bash
root@hostname:/ros2_ws#ros2 bag play -s rosbag_v2 /path/to/rosbagfile.bag

ymd-stella replied:
Rather than replacing it with a version of foxy, simply remove it.

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+FROM ros:galactic-ros1-bridge
+ENV DEBIAN_FRONTEND noninteractive
+
+# ROS
+ENV ROS1_DISTRO noetic
+ENV ROS2_DISTRO galactic
+
```
This may be unnecessary.
See https://github.com/osrf/docker_images/blob/master/ros/galactic/ubuntu/focal/ros1-bridge/Dockerfile#L11-L12.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+    git clone -b wip/approximate-time-rgbd https://github.com/OpenVSLAM-Community/openvslam_ros.git
+
+In case of openvslam using rosbag file, then replace `Dockerfile.socket` with `Dockerfile.socket-rosbag`.
```
It seems that the old description still exists.

Udaykiran87 replied:
Is it also required to delete below lines?

**If you are going to use RGBD setup then exceute following in your terminal.
    git clone -b wip/approximate-time-rgbd https://github.com/OpenVSLAM-Community/openvslam_ros.git**

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+  apt-get autoremove -y -qq && \
+  rm -rf /var/lib/apt/lists/*
+
+WORKDIR /ros2_ws
+
+ARG NUM_THREADS=8
+RUN set -x && \
+  : "build ROS2 packages" && \
+  bash -c "source /opt/ros/${ROS_DISTRO}/setup.bash; \
+  colcon build --parallel-workers ${NUM_THREADS} --cmake-args \
+    -DBUILD_WITH_MARCH_NATIVE=ON \
+    -DUSE_PANGOLIN_VIEWER=OFF \
+    -DUSE_SOCKET_PUBLISHER=ON \
+    -DUSE_STACK_TRACE_LOGGER=ON \
+    -DBOW_FRAMEWORK=DBoW2"
+
```
Not required.

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+    -DBOW_FRAMEWORK=DBoW2"
+
+RUN set -x && \
+  sh -c "echo '#'\!'/bin/bash\nset -e\nsource /opt/ros/${ROS_DISTRO}/setup.bash\nsource /ros2_ws/install/setup.bash\nexec \"\$@\"' > /ros_entrypoint.sh" && \
```
```suggestion
  sh -c "echo '#'\!'/bin/bash\nset -e\nsource /opt/ros/${ROS_DISTRO}/setup.bash\nexec \"\$@\"' > /ros_entrypoint.sh" && \
```

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+
+`Terminal2:`
+Launch the container of OpenVSLAM-socket. The shell interface will be launched in the docker container.Download an ORB vocabulary from GitHub. Publish Images Captured by a USB Camera using package like image_tools.
```
```suggestion

Launch the container of OpenVSLAM-socket. The shell interface will be launched in the docker container. Download an ORB vocabulary from GitHub. Publish Images Captured by a USB Camera using package like image_tools.
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+    xhost +local:
+    docker run --rm -it --name openvslam-ros-socket --device /dev/video0 --net=host openvslam-ros-socket
+    root@hostname:/ros2_ws#curl -sL "https://github.com/OpenVSLAM-Community/FBoW_orb_vocab/raw/main/orb_vocab.fbow" -o orb_vocab.fbow
```
I think it would be hard to read without a space after `#`.

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+    root@hostname:/ros2_ws#ros2 run openvslam_ros run_localization -v ./orb_vocab.fbow -c /path/to/config.yaml --frame-skip 1 --map-db aist_living_lab_1_map.msg
+
+### Step 5.3: Using ROSBAG file
+Execute the following commands:
```
```suggestion
### Step 5.3: Using ROSBAG file

Execute the following commands:
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+`Terminal2:`
+Launch container of openvslam-ros-socket. The shell interface will be launched in the docker container.
+    
+    docker run --rm -it --name openvslam-ros-socket --net=host openvslam-ros-socket   
+OR if you would like to mount a folder (ex:Inputs) to access all required files like config.yaml, rosbag.bag etc then you can run below:
```
```suggestion
    docker run --rm -it --name openvslam-ros-socket --net=host openvslam-ros-socket  
 
OR if you would like to mount a folder (ex:Inputs) to access all required files like config.yaml, rosbag.bag etc then you can run below:
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+Execute the following commands:
+
+    git clone https://github.com/OpenVSLAM-Community/openvslam.git openvslam
+    cp -r openvslam/viewer ./viewer
+    rm -r openvslam/ --force
+    cd viewer
+    docker build -t openvslam-ros-server .
```
```suggestion
    git clone https://github.com/OpenVSLAM-Community/openvslam.git
    cd openvslam/viewer
    docker build -t openvslam-ros-server .
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+
+Execute the following commands:
+
+    cd /path/to/openvslam
+    git submodule update -i --recursive
```
```suggestion
    cd /path/to/openvslam_ros
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on README.md:
```diff
@@ -27,4 +27,265 @@
+For example:
+
+    # building the docker image with four threads
+    docker build -t openvslam-socket -f Dockerfile.socket . --build-arg NUM_THREADS=4
```
```suggestion
    docker build -t openvslam-ros-socket -f Dockerfile.socket . --build-arg NUM_THREADS=4
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+
+# ROS2
+RUN set -x && \
+  sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list' && \
```
This may be unnecessary.
See https://github.com/osrf/docker_images/blob/master/ros/galactic/ubuntu/focal/ros1-bridge/Dockerfile#L6.

Udaykiran87 replied:
I did not delete the above line as it gave below build error.
**Build error:**
```
E: Unable to locate package ros-galactic--image-transport
E: Unable to locate package ros-galactic--cv-bridge
E: Unable to locate package ros-galactic--rosbag2-bag-v2-plugins
E: Unable to locate package ros-galactic--ros2bag
E: Unable to locate package ros-galactic--rosbag2-transport
E: Unable to locate package ros-galactic--rosbag2-storage-default-plugins
```
To solve above, i had to bring back below line:
`sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list' && \`

Now it is working.

Udaykiran87 replied:
@ymd-stella 
Deleted above line as suggested and rechecked, build is through and i am able to use ros2 bag play info command.
Sorry for the confusion.

Uploaded final Dockerfile.rosbag for review.

ymd-stella commented on Dockerfile.rosbag:
```diff
@@ -0,0 +1,48 @@
+    -DUSE_STACK_TRACE_LOGGER=ON \
+    -DBOW_FRAMEWORK=DBoW2"
+
+RUN set -x && \
+  sh -c "echo '#'\!'/bin/bash\nset -e\nsource /opt/ros/${ROS_DISTRO}/setup.bash\nsource /ros2_ws/install/setup.bash\nexec \"\$@\"' > /ros_entrypoint.sh" && \
+  chmod u+x /ros_entrypoint.sh
+
+ENTRYPOINT ["/ros_entrypoint.sh"]
```
The problem mentioned at https://github.com/OpenVSLAM-Community/openvslam_ros/pull/31#discussion_r666297717 is solved by the following change.

```suggestion
ENTRYPOINT []
```

Udaykiran87 replied:
Updated as per review comment.

ymd-stella replied:
> Build error:
> ```
> E: Unable to locate package ros-galactic--image-transport
> E: Unable to locate package ros-galactic--cv-bridge
> E: Unable to locate package ros-galactic--rosbag2-bag-v2-plugins
> E: Unable to locate package ros-galactic--ros2bag
> E: Unable to locate package ros-galactic--rosbag2-transport
> E: Unable to locate package ros-galactic--rosbag2-storage-default-plugins
> ```
> To solve above, i had to bring back below line:
> `sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list' && \`

I can build it without `sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list' && \` in my environment. What is the difference?

ymd-stella replied:
It looks like there is one more hyphen.

Udaykiran87 replied:
Could you please tell me which hyphen you are talking about? I am unable to figure it out.

Udaykiran87 replied:
@ymd-stella 
Deleted above line as suggested and rechecked, build is through and i am able to use ros2 bag play info command.
Sorry for the confusion.

Uploaded final Dockerfile.rosbag for review.
